cpp/ql/test/library-tests/dataflow/taint-tests/bsd.cpp

```diff
@@ -0,0 +1,64 @@
+void sink(...);
+int source();
+
+// --- accept ---
+
+struct sockaddr {
+	unsigned char length;
+	int sa_family;
+	char* sa_data;
+};
+
+int accept(int, const sockaddr*, int*);
+
+void sink(sockaddr);
+
+void test_accept() {
+  int s = source();
+  sockaddr addr;
+  int size = sizeof(sockaddr);
+  int a = accept(s, &addr, &size);
+
+  sink(a); // $ ast=17:11 SPURIOUS: ast=18:12 MISSING: ir
+  sink(addr); // $ ast MISSING: ir
+}
+
+// --- poll ---
+
+struct pollfd {
+  int   fd;
+  short events;
+  short revents;
+};
+
+int poll(struct pollfd *, int, int);
+
+void test_poll() {
+  pollfd pfds;
+
+  pfds.events = 1;
+  pfds.fd = source();
+  poll(&pfds, 1, -1);
+
+  sink(pfds); // $ ast MISSING: ir
+}
+
+// --- select ---
+
+typedef struct {} timeval;
+
+typedef struct fd_set {
+  int  fd_count;
+  int fd_array[4096];
+} fd_set;
+
+int select(int, fd_set *, fd_set *, fd_set *, timeval *);
+
+void test_select(timeval timeout) {
+  fd_set readfds;
+
+  readfds.fd_count = 1;
+  readfds.fd_array[0] = source();
+  select(2, &readfds, nullptr, nullptr, &timeout);
+  sink(&readfds); // $ ast MISSING: ir
```

Comment:
Please could you talk me through what's going on with taint in the `poll` and `select` examples?

Reply:
Sure. In the `poll` case, we taint the file descriptor, and `poll` wait for some I/O event, and then fill the `revents` field with the returned event.

In the `select` case, we taint a file descriptor in `readfds`. [It looks like](https://linux.die.net/man/3/fd_set) this is normally done using the `FD_SET` function which we haven't modeled yet. The call to `select` then taints the `readfds` structure, which we can then use to propagate taint to other `FD_` functions.

Thinking about this taint flow again, I'm actually not sure we want taint to propagate taint to the entire object returned by these functions.

Reply:
50dd0c98e9786275b4696cf7a3f7dfcb64913853 removes the taint from these functions.